controller/formation.go

```diff
@@ -303,7 +318,14 @@ func (r *FormationRepo) unsubscribeLocked(ch chan *ct.ExpandedFormation) {
 		for range ch {
 		}
 	}()
+	si, ok := r.subscriptions[ch]
 	delete(r.subscriptions, ch)
+	if ok {
+		si.uc <- struct{}{}
+		si.wg.Wait()
```

Comment:
But that way we are not ALSO selecting on stopCh inside sendUpdatedSince. Right?

--jmf

> On 25/11/2015, at 02:51, Lewis Marshall <notifications@github.com> wrote:
> 
> In controller/formation.go:
> 
> >  	delete(r.subscriptions, ch)
> > +	if ok {
> > +		si.uc <- struct{}{}
> > +		si.wg.Wait()
> We want the sendUpdatedSince goroutine to stop if the channel is closed, which we can do by just wrapping the channel:
> 
> type FormationSubscription struct {
>     mtx sync.Mutex
>     ch  chan *ct.ExpandedFormation
> }
> 
> func (f *FormationSubscription) Notify(ef *ct.ExpandedFormation) bool {
>     f.mtx.RLock()
>     defer f.mtx.RUnlock()
>     if f.ch == nil {
>         return false
>     }
>     f.ch <- ef
>     return true
> }
> 
> func (f *FormationSubscription) Close() {
>     f.mtx.Lock()
>     defer f.mtx.Unlock()
>     close(f.ch)
>     f.ch = nil
> }
> 
> // in sendUpdatedSince
> for rows.Next() {
>     // ...
>     if ok := f.Notify(ef); !ok {
>         return nil
>     }
> }
> WDYT?
> 
> —
> Reply to this email directly or view it on GitHub.
>